clarity/src/vm/ast/parser_v2/lexer/mod.rs

```diff
@@ -0,0 +1,1957 @@
+                advance = false;
+                self.read_char();
+                if self.next.is_ascii_digit() {
+                    Token::Int(-self.read_integer())
+                } else {
+                    Token::Minus
+                }
+            }
+            '<' => {
+                self.read_char();
+                if self.next == '=' {
+                    Token::LessEqual
+                } else if self.next.is_ascii_alphabetic() {
+                    Token::TraitIdent(self.read_trait_identifier())
+                } else {
+                    advance = false;
+                    Token::Less
+                }
+            }
+            '>' => {
+                self.read_char();
+                if self.next == '=' {
+                    Token::GreaterEqual
+                } else {
+                    advance = false;
+                    Token::Greater
+                }
+            }
+            ';' => {
+                self.read_char();
+                if self.next != ';' {
+                    // If there is just one ';', report an error but continue
+                    //  parsing as if there were two (a comment).
+                    self.add_diagnostic(PlacedError {
+                        span: Span {
+                            start_line: self.last_line as u32,
+                            start_column: self.last_column as u32,
+                            end_line: self.last_line as u32,
+                            end_column: self.last_column as u32,
+                        },
+                        e: LexerError::SingleSemiColon,
+                    });
+                } else {
+                    self.read_char();
+                }
+                advance = false;
+                let comment = self.read_line();
+                Token::Comment(comment.trim_start().to_string())
+            }
+            '\'' => {
+                advance = false;
+                Token::Principal(self.read_principal())
+            }
+            'u' => {
+                advance = false;
+                self.read_char();
+                if self.next.is_ascii_digit() {
+                    Token::Uint(self.read_unsigned())
+                } else if self.next == '"' {
+                    Token::Utf8String(self.read_utf8_string())
+                } else {
+                    Token::Ident(self.read_identifier(Some('u')))
+                }
+            }
+            ' ' | '\t' | '\r' | '\n' => {
+                self.skip_whitespace();
+                advance = false;
+                Token::Whitespace
+            }
+            '"' => {
+                advance = false;
+                Token::AsciiString(self.read_ascii_string())
+            }
+            '0' => {
+                advance = false;
+                self.read_char();
+                if self.next == 'x' {
+                    Token::Bytes(self.read_hex())
+                } else if self.next.is_ascii_digit() {
```

Comment:
Wait, what about the string "0)"?  Such as in `(print (< x 0))`?

Reply:
That would reach the next case,

```rust
                } else if is_separator(self.next) {
                    Token::Int("0".to_string())
                }
```